### PR TITLE
Add backward compat tests

### PR DESCRIPTION
**Pull request by daniellok-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/daniellok-db/mlflow/pull/10629?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10629/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10629
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Add some backward compatibility tests for custom pyfuncs. I logged a model with `mlflow==2.7.1` and `2.8.1` using the following code:

```python
import mlflow

class MyModel(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input):
        return model_input

model = MyModel()
mlflow.pyfunc.save_model(
    python_model=model, 
    path="../tests/resources/pyfunc_models/2.7.1"
)
```

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Verified that it fails when installing `mlflow==2.9.0`:
Screenshot 2023-12-07 at 10 56 44 AM

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/pyfunc/test_backward_compatibility.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import pytest
+
+import mlflow
+
+
+@pytest.mark.parametrize("version", ["2.7.1", "2.8.1"])
+def test_backward_compatibility(version):
+    model = mlflow.pyfunc.load_model(f"tests/resources/pyfunc_models/{version}")
+    assert model.predict("MLflow is great!") == "MLflow is great!"
```
Comment (harupy) [inline tests/pyfunc/test_backward_compatibility.py:9]: /ai I need your help. Please write that that does the following:

1. Create a temporary directory using tmp_path fixture
2. Save the following code as save.py in that directory
3. Create a docker container with python 3.8
4. Install a specified version of mlflow in the container
5. Run the script.
6. Load the model with `mlflow.pyfunc.load_model`
7. Assert the predict reuslt.

```python
save.py

import mlflow

class MyModel(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input):
        return model_input

model = MyModel()
mlflow.pyfunc.save_model(
    python_model=model, 
    path="model"
)
```

Comment (daniellok-db) [reply tests/pyfunc/test_backward_compatibility.py:9]: i think this is essentially what @BenWilson2 is doing in https://github.com/mlflow/mlflow/pull/10626/files#diff-0ff0c4b1146adee8062dbda677190fbf8e68095a1bebaa84e561afa29c75e520R1490

Comment (dbczumar) [reply tests/pyfunc/test_backward_compatibility.py:9]: Not following why this is necessary. It seems sufficient to check in the model files and then try to load them from the current MLflow installation on dev.

Comment (harupy) [reply tests/pyfunc/test_backward_compatibility.py:9]: I agree, let's use locally saved files.

File: tests/resources/pyfunc_models/2.7.1/conda.yaml
[the remark this answers is not in the document]
Comment (daniellok-db) [reply tests/resources/pyfunc_models/2.7.1/conda.yaml:1]: yeah sounds good!

**Review by BenWilson2 — APPROVED** [review]
LGTM once the README is created :)

File: tests/resources/pyfunc_models/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Historical Pyfunc Models
```
Comment (daniellok-db) [inline tests/resources/pyfunc_models/README.md:1]: @harupy is this something like what you had in mind?

Comment (BenWilson2) [reply tests/resources/pyfunc_models/README.md:1]: looks great!

Comment (daniellok-db) [reply tests/resources/pyfunc_models/README.md:1]: awesome, i'll rebase and see if the tests pass

File: tests/resources/pyfunc_models/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+class MyModel(mlflow.pyfunc.PythonModel):
+    def predict(self, context, model_input):
+        return model_input
+
+
+model = MyModel()
+
+mlflow.pyfunc.save_model(
+    python_model=model,
+    path="tests/resources/pyfunc_models/2.8.1",  # replace with the version you installed
```
Comment (harupy) [inline tests/resources/pyfunc_models/README.md:23]:
```suggestion
    path=f"tests/resources/pyfunc_models/{mlflow.__version__}",  # replace with the version you installed
```

Comment (daniellok-db) [reply tests/resources/pyfunc_models/README.md:23]: 🤯

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**